File: Questions/count_primes/count_primes.rs
use std::io;

struct DSA;

impl DSA {
    pub fn count_primes(n: usize) -> usize {
        let mut primes = vec![true; n + 1];
        primes[0] = false;
        primes[1] = false;
        for i in 2..n {
            if primes[i] {
                for j in (i * i..=n).step_by(i) {
                    primes[j] = false;
                }
            }
        }
        primes.iter().filter(|&b| *b == true).count()
    }
}

fn main() {
    let mut input = String::new();
    io::stdin()
        .read_line(&mut input)
        .expect("Failed to read line");
    let n: usize = input.trim().parse().expect("Invalid input");

    println!(
        "There are {} primes less than or equal to {}",
        DSA::count_primes(n),
        n
    );
}

File: Topic-wise/Introduction/count_primes/count_primes.py
from math import isqrt
class DSA:
    def count_primes(self, n):
        primes = [True] * (n+1)
        primes[0] = primes[1] = False
        for idx, val in enumerate(primes[2:], start=2):
            if val:
                for i in range(idx * idx, n+1, idx):
                    primes[i] = False
        return sum(primes)
            

dsa = DSA()
n = int(input())
res = dsa.count_primes(n)
print(f'There are {res} primes less than or equal to {n}')
